File: server/src/utils/judge0Client.js
/**
 * Judge0 API client for code execution
 * This client communicates directly with Judge0 API for code evaluation
 */

const JUDGE0_BASE_URL = process.env.JUDGE0_BASE_URL || "https://judge0-ce.p.rapidapi.com";
const JUDGE0_API_KEY = process.env.JUDGE0_API_KEY || "";

// Language ID mapping for Judge0
const LANGUAGE_IDS = {
    'python': 71,      // Python 3.8.1
    'javascript': 63,  // JavaScript (Node.js 12.14.0)
    'java': 62,        // Java (OpenJDK 13.0.1)
    'cpp': 54,         // C++ (GCC 9.2.0)
    'c': 50,           // C (GCC 9.2.0)
    'csharp': 51,      // C# (Mono 6.6.0.161)
    'go': 60,          // Go (1.13.5)
    'rust': 73,        // Rust (1.40.0)
};

class Judge0Client {
    
    /**
     * Submit code for execution against test cases
     * @param {boolean} waitForResult - If true, wait for execution to complete before returning
     */
    static async submitCode(submissionData, waitForResult = false) {
        try {
            const { 
                submission_id, 
                source_code, 
                language, 
                test_cases, 
                time_limit = 2.0, 
                memory_limit = 256 
            } = submissionData;

            const languageId = LANGUAGE_IDS[language];
            if (!languageId) {
                throw new Error(`Unsupported language: ${language}`);
            }

            console.log(`Executing code with Judge0 for submission ${submission_id}...`);
            console.log(`Language: ${language} (ID: ${languageId}), Test cases: ${test_cases.length}, Wait: ${waitForResult}`);

            const results = [];
            
            // Submit code for each test case
            for (let i = 0; i < test_cases.length; i++) {
                const testCase = test_cases[i];
                
                const submissionPayload = {
                    language_id: languageId,
                    source_code: source_code,
                    stdin: testCase.input || '',
                    expected_output: testCase.expected_output || '',
                    cpu_time_limit: time_limit,
                    memory_limit: memory_limit * 1024, // Convert MB to KB
                };

                try {
                    // Submit for execution
                    const submitResponse = await fetch(`${JUDGE0_BASE_URL}/submissions?base64_encoded=false&wait=false`, {
                        method: 'POST',
                        headers: {
                            'Content-Type': 'application/json',
                            'X-RapidAPI-Key': JUDGE0_API_KEY,
                            'X-RapidAPI-Host': 'judge0-ce.p.rapidapi.com'
                        },
                        body: JSON.stringify(submissionPayload)
                    });

                    if (!submitResponse.ok) {
                        console.error(`Judge0 submission failed for test case ${i}:`, await submitResponse.text());
                        results.push({
                            test_case_index: i,
                            input: testCase.input,
                            expected_output: testCase.expected_output,
                            passed: false,
                            status_id: 13, // Internal Error
                            stdout: '',
                            stderr: 'Submission failed',
                            time: 0,
                            memory: 0
                        });
                        continue;
                    }

                    const submission = await submitResponse.json();
                    const token = submission.token;

                    if (!token) {
                        console.error('No token received from Judge0');
                        results.push({
                            test_case_index: i,
                            input: testCase.input,
                            expected_output: testCase.expected_output,
                            passed: false,
                            status_id: 13,
                            stdout: '',
                            stderr: 'No token received',
                            time: 0,
                            memory: 0
                        });
                        continue;
                    }

                    // Wait for execution to complete and get result
                    const result = await this._getSubmissionResult(token);
                    result.test_case_index = i;
                    result.input = testCase.input;
                    result.expected_output = testCase.expected_output;
                    
                    // Check if output matches expected
                    const stdout = (result.stdout || '').trim();
                    const expected = (testCase.expected_output || '').trim();
                    result.passed = result.status_id === 3 && stdout === expected;
                    
                    results.push(result);

                } catch (error) {
                    console.error(`Error executing test case ${i}:`, error);
                    results.push({
                        test_case_index: i,
                        input: testCase.input,
                        expected_output: testCase.expected_output,
                        passed: false,
                        status_id: 13,
                        stdout: '',
                        stderr: error.message,
                        time: 0,
                        memory: 0
                    });
                }
            }

            // Analyze overall results
            const analysisResult = this._analyzeResults(results);
            
            console.log(`Execution completed for submission ${submission_id}:`, {
                status: analysisResult.status,
                passed: analysisResult.passed_tests,
                total: analysisResult.total_tests
            });

            // If this is a Run (waitForResult=true), return immediately
            if (waitForResult) {
                return analysisResult;
            }

            // For Submit, send results back to submission controller asynchronously
            const backendUrl = process.env.BACKEND_URL || 'http://localhost:5000';
            await fetch(`${backendUrl}/api/v1/submissions/${submission_id}/result`, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(analysisResult)
            }).catch(err => {
                console.error('Failed to update submission result:', err);
            });

            return analysisResult;

        } catch (error) {
            console.error('Judge0 submission error:', error);
            throw error;
        }
    }

    /**
     * Get submission result by token (with retries)
     */
    static async _getSubmissionResult(token, maxAttempts = 20) {  // Increased from 10 to 20
        console.log(`Polling Judge0 for token: ${token} (max attempts: ${maxAttempts})`);
        
        for (let attempt = 0; attempt < maxAttempts; attempt++) {
            try {
                const response = await fetch(`${JUDGE0_BASE_URL}/submissions/${token}?base64_encoded=false&fields=*`, {
                    method: 'GET',
                    headers: {
                        'X-RapidAPI-Key': JUDGE0_API_KEY,
                        'X-RapidAPI-Host': 'judge0-ce.p.rapidapi.com'
                    }
                });

                if (response.ok) {
                    const result = await response.json();
                    const status = result.status || {};
                    const statusId = status.id;

                    // Log current status for debugging
                    if (attempt % 5 === 0 || statusId !== 1 && statusId !== 2) {
                        console.log(`Attempt ${attempt + 1}/${maxAttempts}: Status ${statusId} (${status.description})`);
                    }

                    // Status IDs: 1 = In Queue, 2 = Processing
                    // If not in queue or processing, execution is complete
                    if (statusId !== 1 && statusId !== 2) {
                        console.log(`✓ Execution completed: ${status.description}`);
                        return {
                            status_id: statusId,
                            status_description: status.description || 'Unknown',
                            stdout: result.stdout || '',
                            stderr: result.stderr || '',
                            compile_output: result.compile_output || '',
                            time: parseFloat(result.time) || 0,
                            memory: parseInt(result.memory) || 0
                        };
                    }
                }

                // Wait before next attempt (reduced from 1000ms to 500ms for faster polling)
                await new Promise(resolve => setTimeout(resolve, 500));

            } catch (error) {
                console.error(`Error getting submission result (attempt ${attempt + 1}):`, error.message);
                await new Promise(resolve => setTimeout(resolve, 500));
            }
        }

        // Timeout after max attempts
        console.error(`✗ Timeout after ${maxAttempts} attempts for token: ${token}`);
        return {
            status_id: 13, // Internal Error
            status_description: 'Execution timeout',
            stdout: '',
            stderr: 'Maximum retry attempts reached. Judge0 API may be overloaded or code execution is taking too long.',
            compile_output: '',
            time: 0,
            memory: 0
        };
    }

    /**
     * Analyze execution results across all test cases
     */
    static _analyzeResults(results) {
        if (!results || results.length === 0) {
            return {
                status: 'runtime_error',
                error_message: 'No results',
                execution_time: 0,
                memory_usage: 0,
                passed_tests: 0,
                total_tests: 0,
                test_results: []
            };
        }

        let passedTests = 0;
        let totalExecutionTime = 0;
        let maxMemory = 0;
        let firstError = null;

        for (const result of results) {
            const statusId = result.status_id || 13;

            // Check if test passed
            if (result.passed) {
                passedTests++;
            }

            // Track performance metrics
            totalExecutionTime += result.time || 0;
            maxMemory = Math.max(maxMemory, result.memory || 0);

            // Capture first error
            if (!result.passed && !firstError) {
                firstError = {
                    status_id: statusId,
                    description: result.status_description || 'Unknown error',
                    stderr: result.stderr || '',
                    compile_output: result.compile_output || ''
                };
            }
        }

        // Determine overall status
        const overallStatus = this._determineOverallStatus(results, passedTests, results.length);

        return {
            status: overallStatus,
            error_message: this._getErrorMessage(firstError),
            execution_time: totalExecutionTime,
            memory_usage: maxMemory,
            passed_tests: passedTests,
            total_tests: results.length,
            test_results: results.map(r => ({
                input: r.input,
                expectedOutput: r.expected_output,
                actualOutput: r.stdout || '',
                passed: r.passed,
                status: r.status_description,
                time: r.time,
                memory: r.memory
            }))
        };
    }

    /**
     * Determine overall submission status
     */
    static _determineOverallStatus(results, passed, total) {
        if (passed === total) {
            return 'accepted';
        }

        // Check for common error types
        for (const result of results) {
            const statusId = result.status_id || 13;

            if (statusId === 5) return 'time_limit_exceeded';
            if (statusId === 6) return 'compilation_error';
            if (statusId >= 7 && statusId <= 12) return 'runtime_error';
            if (statusId === 17) return 'memory_limit_exceeded';
        }

        // Default to wrong answer
        return 'wrong_answer';
    }

    /**
     * Get human-readable error message
     */
    static _getErrorMessage(error) {
        if (!error) return '';

        const description = error.description || '';
        const stderr = (error.stderr || '').trim();
        const compileOutput = (error.compile_output || '').trim();

        if (compileOutput) {
            return `Compilation Error: ${compileOutput}`;
        } else if (stderr) {
            return `${description}: ${stderr}`;
        } else {
            return description;
        }
    }

    /**
     * Get execution result by submission ID (legacy method for compatibility)
     */
    static async getResult(submissionId) {
        try {
            const backendUrl = process.env.BACKEND_URL || 'http://localhost:5000';
            const response = await fetch(`${backendUrl}/api/v1/submissions/${submissionId}`);

            if (!response.ok) {
                throw new Error(`Failed to get result: ${response.statusText}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Error getting submission result:', error);
            throw new Error('Failed to retrieve submission result');
        }
    }

    /**
     * Generate problem using AI service (forwarding to Python service)
     */
    static async generateProblem(problemData) {
        try {
            const JUDGE0_SERVICE_URL = process.env.JUDGE0_SERVICE_URL || "http://localhost:8000";
            const response = await fetch(`${JUDGE0_SERVICE_URL}/api/v1/problems/generate`, {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify(problemData)
            });

            if (!response.ok) {
                throw new Error(`Problem generation service error: ${response.statusText}`);
            }

            return await response.json();
        } catch (error) {
            console.error('Error communicating with problem generation service:', error);
            throw new Error('Problem generation service unavailable');
        }
    }
}

export default Judge0Client;

File: service/app/services/judge0_client.py
"""Judge0 API client for code execution."""

import httpx
import asyncio
import logging
from typing import Dict, Any, Optional, List
from app.core.config import settings
from app.schemas.schemas import ProgrammingLanguage, SubmissionStatus

logger = logging.getLogger(__name__)


class Judge0Client:
    """Client for Judge0 API integration."""
    
    # Language ID mapping for Judge0
    LANGUAGE_IDS = {
        ProgrammingLanguage.PYTHON: 71,      # Python 3.8.1
        ProgrammingLanguage.JAVASCRIPT: 63,  # JavaScript (Node.js 12.14.0)
        ProgrammingLanguage.JAVA: 62,        # Java (OpenJDK 13.0.1)
        ProgrammingLanguage.CPP: 54,         # C++ (GCC 9.2.0)
        ProgrammingLanguage.C: 50,           # C (GCC 9.2.0)
        ProgrammingLanguage.CSHARP: 51,      # C# (Mono 6.6.0.161)
        ProgrammingLanguage.GO: 60,          # Go (1.13.5)
        ProgrammingLanguage.RUST: 73,        # Rust (1.40.0)
    }
    
    def __init__(self):
        self.base_url = settings.JUDGE0_BASE_URL
        self.headers = {
            "X-RapidAPI-Key": settings.JUDGE0_API_KEY,
            "X-RapidAPI-Host": "judge0-ce.p.rapidapi.com",
            "Content-Type": "application/json"
        }
    
    async def submit_code(
        self,
        source_code: str,
        language: ProgrammingLanguage,
        test_cases: List[Dict[str, str]],
        time_limit: Optional[float] = None,
        memory_limit: Optional[int] = None
    ) -> Dict[str, Any]:
        """Submit code for execution against test cases."""
        try:
            language_id = self.LANGUAGE_IDS.get(language)
            if not language_id:
                raise ValueError(f"Unsupported language: {language}")
            
            results = []
            
            async with httpx.AsyncClient(timeout=30.0) as client:
                # Submit code for each test case
                for i, test_case in enumerate(test_cases):
                    submission_data = {
                        "language_id": language_id,
                        "source_code": source_code,
                        "stdin": test_case["input"],
                        "expected_output": test_case["expected_output"],
                        "cpu_time_limit": time_limit or 2.0,
                        "memory_limit": (memory_limit or 256) * 1024  # Convert MB to KB
                    }
                    
                    # Submit for execution
                    response = await client.post(
                        f"{self.base_url}/submissions",
                        json=submission_data,
                        headers=self.headers,
                        params={"base64_encoded": "false", "wait": "false"}
                    )
                    
                    if response.status_code != 201:
                        logger.error(f"Submission failed: {response.text}")
                        continue
                    
                    submission = response.json()
                    token = submission.get("token")
                    
                    if token:
                        # Wait for execution and get result
                        result = await self._get_submission_result(client, token)
                        result["test_case_index"] = i
                        result["input"] = test_case["input"]
                        result["expected_output"] = test_case["expected_output"]
                        results.append(result)
            
            # Analyze overall results
            return self._analyze_results(results)
            
        except Exception as e:
            logger.error(f"Judge0 submission error: {e}")
            return {
                "status": SubmissionStatus.RUNTIME_ERROR,
                "error_message": str(e),
                "execution_time": 0,
                "memory_usage": 0,
                "passed_tests": 0,
                "total_tests": len(test_cases),
                "test_results": []
            }
    
    async def _get_submission_result(self, client: httpx.AsyncClient, token: str) -> Dict[str, Any]:
        """Get submission result by token."""
        max_attempts = 10
        attempt = 0
        
        while attempt < max_attempts:
            try:
                response = await client.get(
                    f"{self.base_url}/submissions/{token}",
                    headers=self.headers,
                    params={"base64_encoded": "false", "fields": "*"}
                )
                
                if response.status_code == 200:
                    result = response.json()
                    
                    # Check if execution is complete
                    if result.get("status", {}).get("id") not in [1, 2]:  # Not "In Queue" or "Processing"
                        return self._parse_judge0_result(result)
                
                # Wait before next attempt
                await asyncio.sleep(1)
                attempt += 1
                
            except Exception as e:
                logger.error(f"Error getting submission result: {e}")
                attempt += 1
                await asyncio.sleep(1)
        
        return {
            "status_id": 13,  # Internal Error
            "stdout": "",
            "stderr": "Execution timeout",
            "time": 0,
            "memory": 0
        }
    
    def _parse_judge0_result(self, judge0_result: Dict[str, Any]) -> Dict[str, Any]:
        """Parse Judge0 API result."""
        status = judge0_result.get("status", {})
        status_id = status.get("id", 13)
        
        return {
            "status_id": status_id,
            "status_description": status.get("description", "Unknown"),
            "stdout": judge0_result.get("stdout", ""),
            "stderr": judge0_result.get("stderr", ""),
            "compile_output": judge0_result.get("compile_output", ""),
            "time": float(judge0_result.get("time") or 0),
            "memory": int(judge0_result.get("memory") or 0)
        }
    
    def _analyze_results(self, results: List[Dict[str, Any]]) -> Dict[str, Any]:
        """Analyze execution results across all test cases."""
        if not results:
            return {
                "status": SubmissionStatus.RUNTIME_ERROR,
                "error_message": "No results",
                "execution_time": 0,
                "memory_usage": 0,
                "passed_tests": 0,
                "total_tests": 0,
                "test_results": []
            }
        
        passed_tests = 0
        total_execution_time = 0
        max_memory = 0
        first_error = None
        
        for result in results:
            status_id = result.get("status_id", 13)
            
            # Check if test passed
            if status_id == 3:  # Accepted
                stdout = result.get("stdout", "").strip()
                expected = result.get("expected_output", "").strip()
                if stdout == expected:
                    passed_tests += 1
            
            # Track performance metrics
            total_execution_time += result.get("time", 0)
            max_memory = max(max_memory, result.get("memory", 0))
            
            # Capture first error
            if status_id != 3 and not first_error:
                first_error = {
                    "status_id": status_id,
                    "description": result.get("status_description", "Unknown error"),
                    "stderr": result.get("stderr", ""),
                    "compile_output": result.get("compile_output", "")
                }
        
        # Determine overall status
        overall_status = self._determine_overall_status(results, passed_tests, len(results))
        
        return {
            "status": overall_status,
            "error_message": self._get_error_message(first_error) if first_error else "",
            "execution_time": total_execution_time,
            "memory_usage": max_memory,
            "passed_tests": passed_tests,
            "total_tests": len(results),
            "test_results": results
        }
    
    def _determine_overall_status(self, results: List[Dict[str, Any]], passed: int, total: int) -> SubmissionStatus:
        """Determine overall submission status."""
        if passed == total:
            return SubmissionStatus.ACCEPTED
        
        # Check for common error types
        for result in results:
            status_id = result.get("status_id", 13)
            
            if status_id == 5:  # Time Limit Exceeded
                return SubmissionStatus.TIME_LIMIT_EXCEEDED
            elif status_id == 6:  # Compilation Error
                return SubmissionStatus.COMPILATION_ERROR
            elif status_id == 7:  # Runtime Error (SIGSEGV)
                return SubmissionStatus.RUNTIME_ERROR
            elif status_id == 8:  # Runtime Error (SIGXFSZ)
                return SubmissionStatus.RUNTIME_ERROR
            elif status_id == 9:  # Runtime Error (SIGFPE)
                return SubmissionStatus.RUNTIME_ERROR
            elif status_id == 10:  # Runtime Error (SIGABRT)
                return SubmissionStatus.RUNTIME_ERROR
            elif status_id == 11:  # Runtime Error (NZEC)
                return SubmissionStatus.RUNTIME_ERROR
            elif status_id == 12:  # Runtime Error (Other)
                return SubmissionStatus.RUNTIME_ERROR
            elif status_id == 17:  # Memory Limit Exceeded
                return SubmissionStatus.MEMORY_LIMIT_EXCEEDED
        
        # Default to wrong answer if no specific error
        return SubmissionStatus.WRONG_ANSWER
    
    def _get_error_message(self, error: Dict[str, Any]) -> str:
        """Get human-readable error message."""
        if not error:
            return ""
        
        description = error.get("description", "")
        stderr = error.get("stderr", "").strip()
        compile_output = error.get("compile_output", "").strip()
        
        if compile_output:
            return f"Compilation Error: {compile_output}"
        elif stderr:
            return f"{description}: {stderr}"
        else:
            return description


# Global Judge0 client instance
judge0_client = Judge0Client()